security/pkg/server/ca/node_auth.go

```diff
@@ -16,27 +16,157 @@ package ca
+		ztunnelPodsClient:        map[cluster.ID]kclient.Client[*v1.Pod]{},
+		filter:                   filter,
+		trustedNodeAccounts:      trustedNodeAccounts,
+	}
+	if addClusterHandler != nil {
+		addClusterHandler(m)
+	}
+	return m
+}
+
+func (mNa *MulticlusterNodeAuthorizor) ClusterAdded(cluster *multicluster.Cluster, _ <-chan struct{}) {
+	na, err := NewClusterNodeAuthorizer(cluster.Client, mNa.filter, mNa.trustedNodeAccounts)
+	if err != nil {
+		serverCaLog.Errorf("failed to initialize node authorizer for cluster %v: %v", cluster.ID, err)
+		return
+	}
+	mNa.m.Lock()
+	defer mNa.m.Unlock()
+	mNa.addCluster(cluster.ID, cluster.Client, na)
+}
+
+func (mNa *MulticlusterNodeAuthorizor) ClusterUpdated(cluster *multicluster.Cluster, stop <-chan struct{}) {
+	na, err := NewClusterNodeAuthorizer(cluster.Client, mNa.filter, mNa.trustedNodeAccounts)
+	if err != nil {
+		serverCaLog.Errorf("failed to initialize node authorizer for cluster %v: %v", cluster.ID, err)
+	}
+	mNa.m.Lock()
+	defer mNa.m.Unlock()
+	mNa.deleteCluster(cluster.ID)
+	mNa.addCluster(cluster.ID, cluster.Client, na)
+}
+
+func (mNa *MulticlusterNodeAuthorizor) ClusterDeleted(key cluster.ID) {
+	mNa.m.Lock()
+	defer mNa.m.Unlock()
+	mNa.deleteCluster(key)
+}
+
+func (mNa *MulticlusterNodeAuthorizor) addCluster(clusterID cluster.ID, client kube.Client, na *ClusterNodeAuthorizer) {
+	ztunnelPodClient := kclient.NewFiltered[*v1.Pod](client, kclient.Filter{
+		ObjectFilter:    isZtunnelPod,
+		ObjectTransform: kube.StripPodUnusedFields,
+	})
+	ztunnelPodClient.AddEventHandler(controllers.EventHandler[*v1.Pod]{
+		AddFunc: func(po *v1.Pod) {
+			serverCaLog.Debugf("pod %v/%v added to ztunnel index, UID: %v, clusteID: %v", po.Namespace, po.Name, po.UID, clusterID)
+			mNa.ztunnelIndex.Store(po.UID, clusterID)
+		},
+		DeleteFunc: func(po *v1.Pod) {
+			serverCaLog.Debugf("pod %v/%v deleted from ztunnel index, UID: %v", po.Namespace, po.Name, po.UID)
+			mNa.ztunnelIndex.Delete(po.UID)
+		},
+	})
+	mNa.ztunnelPodsClient[clusterID] = ztunnelPodClient
+	mNa.remoteNodeAuthenticators[clusterID] = na
+}
+
+func (mNa *MulticlusterNodeAuthorizor) deleteCluster(clusterID cluster.ID) {
+	if _, found := mNa.ztunnelPodsClient[clusterID]; found {
+		serverCaLog.Debugf(" shutting down ztunnel index for cluster %v", clusterID)
+		mNa.ztunnelPodsClient[clusterID].ShutdownHandlers()
+	}
+	mNa.ztunnelIndex.Range(func(key, value any) bool {
+		v := value.(cluster.ID)
+		if v == clusterID {
+			serverCaLog.Debugf("deleting ztunnel index entry for ztunnel pod: %v, cluster %v", key, clusterID)
+			mNa.ztunnelIndex.Delete(key)
+		}
+		return true
+	})
+	delete(mNa.ztunnelPodsClient, clusterID)
+	delete(mNa.remoteNodeAuthenticators, clusterID)
+}
+
+func (mNa *MulticlusterNodeAuthorizor) authenticateImpersonation(caller security.KubernetesInfo, requestedIdentityString string) error {
+	clusterID, found := mNa.ztunnelIndex.Load(types.UID(caller.PodUID))
```

Comment:
we can retrieve the clusterID from `ctx`. Please refer `extractClusterID`

Reply:
Thanks for the valuable suggestion! We now change to reuse `ExtractClusterID` and retrieve the `clusterID` from `ctx`. In this way, there's no need to maintain a `ztunnelIndex` anymore (this index's purpose is just retrieving `clusterID` ). It is a simpler and more reliable implementation.